src/lib/statistics/statistics_objects/top_k_as_generic_histogram.cpp

```diff
@@ -0,0 +1,129 @@
+  for (auto top_k_index = 0u; top_k_index < k; top_k_index++) {
+    auto value_distribution_it = remove(value_distribution.begin(), value_distribution.end(),
+                                        std::make_pair(top_k_names[top_k_index], top_k_counts[top_k_index]));
+    value_distribution.erase(value_distribution_it, value_distribution.end());
+  }
+
+  // Each Top K value is modeled as one bin with height as its stored count.
+  // Between two Top K value bins, one bin is created for the Non-Top K values between them.
+  // Together with a potential last bin after the last Top K value we have a maximum bin count of 2 * K + 1.
+  // If there are no more values stored in value_distribution after the Top K values have been removed,
+  // we only have Top K values and therefore need exactly k bins.
+  const auto bin_count = value_distribution.size() < 1 ? BinID{k} : BinID{2 * top_k_names.size() + 1};
+
+  GenericHistogramBuilder<T> builder{bin_count, domain};
+
+  // Calculate estimate for occurrence count for Non-Top K values assuming a uniform distribution.
+  const auto non_top_k_count =
+      std::accumulate(value_distribution.cbegin(), value_distribution.cend(), HistogramCountType{0},
+                      [](HistogramCountType current_count, const std::pair<T, HistogramCountType>& value_count) {
+                        return current_count + value_count.second;
+                      });
+
+  const auto bin_distinct_count = value_distribution.size();
+  const auto count_per_non_top_k_value = bin_distinct_count != 0 ? non_top_k_count / bin_distinct_count : BinID{0};
+
+  // Construct Generic Histogram with single bins for Top K Values.
+  // For Non-Top K Values, one bin is created for all Non-Top K values between two Top K bins,
+  // using the calculated estimation of count_per_non_top_k_value.
+
+  auto current_minimum_index = 0u;
+  auto current_maximum_index = value_distribution.size() - 1;
+
+  for (auto top_k_index = 0ul, top_k_size = top_k_names.size(); top_k_index < top_k_size; top_k_index++) {
+    const auto current_top_k_value = top_k_names[top_k_index];
+
+    auto value_dist_lower_bound = std::lower_bound(
+        value_distribution.begin() + current_minimum_index, value_distribution.end(), current_top_k_value,
+        [](const auto value_count_pair, auto value) { return value_count_pair.first < value; });
+
+    // For each Top K value a Non-Top K values bin between the previous Top K value and itself,
+    // as well as a Top K value bin are created.
+
+    // We can skip creating a Non-Top K value bin,
+    // if there are no Non-Top K values between the previous and the current Top K value.
+    if (!(value_dist_lower_bound == value_distribution.begin() ||
+          std::prev(value_dist_lower_bound) - value_distribution.begin() < current_minimum_index)) {
+      // Create Non-Top K values bin
+      current_maximum_index = std::prev(value_dist_lower_bound) - value_distribution.begin();
```

Comment:
You could assign `std::prev(value_dist_lower_bound) - value_distribution.begin()` to a variable before line 99 so you can use this variable in the if and here.